undefined reference to 'pthread_yield' when compiling from source with gcc 13.3.0 on Ubuntu 24.04.1 LTS

Undefined reference to 'pthread_yield' issue when compiling spectre from source with gcc 13.3.0 on Ubuntu 24.04.1 LTS

The error message during make included:

" ../lib/libEvents.so: error: undefined reference to 'pthread_yield' "

I tried to bug hunt this with Nils Vu, and it seems that 'pthread_yield' has been replaced with 'sched_yield', and speculated that this issue might be from charm++.

I apologize for the long delay in response. I re-compiled charm++ and spectre with clang rather than gcc to see if that corrected things. Same issue. I have uploaded the BuildInfo.txt file, and here is the error message:

[100%] Linking CXX executable ../../../../../bin/SolveScalarSelfForce
../../../../../lib/libDataStructures.so: error: undefined reference to 'pthread_yield'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/build.make:191: bin/SolveScalarSelfForce] Error 1
make[2]: *** [CMakeFiles/Makefile2:16249: src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:16256: src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/rule] Error 2
make: *** [Makefile:1116: SolveScalarSelfForce] Error 2

[BuildInfo.txt](https://github.com/user-attachments/files/24437282/BuildInfo.txt)

Turning off shared libraries was good advice, it did not immediately work but it revealed this error message:

[100%] Linking CXX executable ../../../../../bin/SolveScalarSelfForce
../../../../../lib/libxsmm.a(libxsmm_gemm.c.o):libxsmm_gemm.c:function libxsmm_mmbatch_kernel:(.text+0x29731): error: undefined reference to 'pthread_yield'
../../../../../lib/libxsmm.a(libxsmm_gemm.c.o):libxsmm_gemm.c:function libxsmm_mmbatch_kernel:(.text+0x29a11): error: undefined reference to 'pthread_yield'
../../../../../lib/libxsmm.a(libxsmm_gemm.c.o):libxsmm_gemm.c:function libxsmm_mmbatch_kernel:(.text+0x29c51): error: undefined reference to 'pthread_yield'
../../../../../lib/libxsmm.a(libxsmm_gemm.c.o):libxsmm_gemm.c:function libxsmm_mmbatch_kernel:(.text+0x2a171): error: undefined reference to 'pthread_yield'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/build.make:218: bin/SolveScalarSelfForce] Error 1
make[2]: *** [CMakeFiles/Makefile2:16334: src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:16341: src/Elliptic/Executables/SelfForce/Scalar/CMakeFiles/SolveScalarSelfForce.dir/rule] Error 2
make: *** [Makefile:1116: SolveScalarSelfForce] Error 2

Which led me to suspect libxsmm. I tried changing to the newest release version of libxsmm, v1.17, but that didn't work either. So I tried cloning the main branch of libxsmm to get the latest stable version, and it worked! Compiled and runs successfully. Thanks for the advice!

--- a/cmake/SetupLIBXSMM.cmake
+++ b/cmake/SetupLIBXSMM.cmake
@@ -14,19 +14,11 @@ if (NOT LIBXSMM_FOUND)
   # https://libxsmm.readthedocs.io/en/latest/#rules-for-building-libxsmm
   include(FetchContent)
   # Need an unreleased version for Apple Silicon chips
-  if (APPLE AND ("${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "arm64"
-                OR "${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "aarch64"))
-    FetchContent_Declare(xsmm
-      GIT_REPOSITORY https://github.com/libxsmm/libxsmm.git
-      GIT_TAG 939f11042fc9ae4bbe975cedb2330d4f9f4bb26e
-      ${SPECTRE_FETCHCONTENT_BASE_ARGS}
-    )
-  else()
-    FetchContent_Declare(xsmm
-      URL https://github.com/libxsmm/libxsmm/archive/1.16.1.tar.gz
-      ${SPECTRE_FETCHCONTENT_BASE_ARGS}
-    )
-  endif()
+  FetchContent_Declare(xsmm
+    GIT_REPOSITORY https://github.com/libxsmm/libxsmm.git
+    GIT_TAG 939f11042fc9ae4bbe975cedb2330d4f9f4bb26e
+    ${SPECTRE_FETCHCONTENT_BASE_ARGS}
+  )
   FetchContent_GetProperties(xsmm)
   if(NOT xsmm_POPULATED)
     FetchContent_Populate(xsmm)

--- a/docs/Installation/Installation.md
+++ b/docs/Installation/Installation.md
@@ -222,8 +222,12 @@ The following dependencies will be fetched automatically if you set
   build and installation (as detailed in the [Catch2
   docs](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md#installing-catch2-from-git-repository)).
   Compile with `CMAKE_POSITION_INDEPENDENT_CODE=ON`.
-* [LIBXSMM](https://github.com/libxsmm/libxsmm) version 1.16.1 or later.
-  \cite Libxsmm
+* [LIBXSMM](https://github.com/libxsmm/libxsmm) version 1.16.1 or later.  Some
+  configurations of LIBXSMM build incorrectly against pthreads, causing
+  ``undefined reference to 'pthread_yield'`` when building SpECTRE.  It is
+  unclear what conditions trigger the problem, but it is fixed in versions
+  newer than 1.17.  (As of this writing, there are no newer releases and
+  affected systems must build from git.) \cite Libxsmm
 * [yaml-cpp](https://github.com/jbeder/yaml-cpp) version 0.7.0 or later.
   Building with shared library support is recommended when installing from
   source. \cite Yamlcpp

